File: AddressSimilarity.py
import math
import re

class addressSimilarity:

    def __init__(self, CHS_COS_WEIGHT = 0.5, CHS_COM_WEIGHT = 0.5, CHS_THRESHOLD = 2, ENG_COS_WEIGHT = 0.3, ENG_COM_WEIGHT = 0.7, ENG_THRESHOLD = 3):
        self.CHS_COS_WEIGHT = CHS_COS_WEIGHT
        self.CHS_COM_WEIGHT = CHS_COM_WEIGHT
        self.CHS_THRESHOLD = CHS_THRESHOLD
        self.ENG_COS_WEIGHT = ENG_COS_WEIGHT
        self.ENG_COM_WEIGHT = ENG_COM_WEIGHT
        self.ENG_THRESHOLD = ENG_THRESHOLD
            
    def arabic2Chinese_full(self, num):
        unit = ['', '十', '百', '千', '萬', '十萬', '百萬', '千萬', '億', '十億', '百億', '千億', '萬億']
        arabicNum = {'0':'零', '1':'一', '2':'二', '3':'三', '4':'四', '5':'五',
            '6':'六', '7':'七', '8':'八', '9':'九', '10':'十'}
        strNum = str(num)
        if num < 0 or num > 9999999999999:
            return num
        if num <= 10:
            return arabicNum[strNum]
        numLen = len(strNum)
        chineseNum = ''
        for i, numChar in enumerate(strNum):
            if numChar == '0' and strNum[i - 1] != '0' and i < len(strNum) - 1:
                chineseNum += arabicNum[numChar]
            elif numChar != '0':
                chineseNum += arabicNum[numChar]
                if not set(unit[numLen - 1]) & set(unit[numLen - 2]):
                    chineseNum += unit[numLen - 1]
                else:
                    if strNum[i + 1] == '0':
                        chineseNum += (unit[numLen - 1])
                    else:
                        chineseNum += (unit[numLen - 1])[0]
            numLen -= 1
        if (len(strNum) > 5 or len(strNum) == 2) and strNum[0] == '1':
            chineseNum = chineseNum[1:]
        return chineseNum[:-1] if num % 100 == 0 else chineseNum

    def arabic2Chinese_simple(self, num):
        arabicNum = {'0':'零', '1':'一', '2':'二', '3':'三', '4':'四', '5':'五',
            '6':'六', '7':'七', '8':'八', '9':'九', '10':'十'}
        chineseNum = ''
        for numChar in str(num):
            chineseNum += arabicNum[numChar]
        return chineseNum
      
    def preProcessing(self, rawAddress1, rawAddress2, escapeChars = [], replaceChars = {}):
        address1 = rawAddress1.strip().lower()
        address2 = rawAddress2.strip().lower()
        for escapeChar in escapeChars:
            address1 = address1.replace(escapeChar, '')
            address2 = address2.replace(escapeChar, '')
        for replaceChar in replaceChars:
            address1 = address1.replace(replaceChar, replaceChars[replaceChar])
            address2 = address2.replace(replaceChar, replaceChars[replaceChar])
        return [address1, address2]
        
    def cosSimilarity(self, address1, address2):
        if address1 == address2:
            return 1.0
        addr1Len, addr2Len = len(address1), len(address2)
        addr1Set, addr2Set = set(address1), set(address2)
        addrSet = addr1Set.union(addr2Set)
        addr1Count, addr2Count = [], []
        crossProduct, addr1Product, addr2Product = [], [], []
        for setChar in addrSet:
            i, j = 0, 0
            for addrChar in address1:
                if setChar == addrChar:
                    i += 1
            addr1Count.append(i)
            for addrChar in address2:
                if setChar == addrChar:
                    j += 1
            addr2Count.append(j)
        for i in range(len(addr1Count)):
            crossProduct.append(addr1Count[i] * addr2Count[i])
            addr1Product.append(addr1Count[i] * addr1Count[i])
            addr2Product.append(addr2Count[i] * addr2Count[i])
        return sum(crossProduct) / (math.sqrt(sum(addr1Product)) * math.sqrt(sum(addr2Product)))
       
    def comSimilarity(self, address1, address2, threshold):
        if ((address1.isalpha() and address2.isalpha()) or (address1.isdigit() and address2.isdigit())) and address1 == address2:
            return 1.0
        addr1Len, addr2Len = len(address1), len(address2)
        strMatrix = [[0] * addr2Len for _ in range(addr1Len)]
        for i in range(addr1Len):
            for j in range(addr2Len):
                if address1[i] == address2[j]:
                    if i == 0 or j == 0:
                        strMatrix[i][j] = 1
                        continue
                    strMatrix[i][j] = strMatrix[i - 1][j - 1] + 1
        score = 0
        i = addr1Len - 1 
        previous = 0 if threshold == 1 else 1
        while i > 0:
            j = addr2Len - 1
            while j > 0:
                if strMatrix[i][j] >= threshold and strMatrix[i - 1][j - 1] >= previous:
                    score += strMatrix[i][j]
                    i -= strMatrix[i][j] - 1
                    break
                j -= 1 
            i -= 1 
        return score / max(addr1Len, addr2Len)
        
    def compareChsAddress(self, rawAddress1, rawAddress2):
        escapeChars = [' ', '　', ',', '，', '.', '。',
            '號', '号', '鋪', '铺', '舖', '九龍', '九龙', '香港', '新界',
            '(', ')', '（', '）', '[', ']', '〔', '〕', '{', '}', '｛', '｝', '|', '｜', '+', '＋',
            '/', '／', '\\', '＼', '~', '～', ';', '；', '*', '＊', '@', '＠', '\'', '’', '"', '”',
            '#', '＃', '!', '！', '$', '＄', '%', '％', '<', '＜', '>', '＞', '^', '︿', '=', '＝', '、']
        replaceChars = {'-':'至', '—':'至', 'o':'零', 'ｏ':'零', 'Ｏ':'零', '&':'及', '＆':'及'}
        address1 = self.preProcessing(rawAddress1, rawAddress2, escapeChars, replaceChars)[0]
        address2 = self.preProcessing(rawAddress1, rawAddress2, escapeChars, replaceChars)[1]
        
        def toChineseNum(address1, address2):
            for num in re.findall('\d+', address1):
                chineseNum1 = self.arabic2Chinese_full(int(num))
                chineseNum2 = self.arabic2Chinese_simple(int(num))
                address1 = address1.replace(num, chineseNum1) if chineseNum1 in address2 else address1.replace(num, chineseNum2)
            return address1
        address1 = toChineseNum(address1, address2)
        address2 = toChineseNum(address2, address1)
        
        cosScore = self.cosSimilarity(address1, address2)
        commonScore = self.comSimilarity(address1, address2, self.CHS_THRESHOLD)

        return cosScore * self.CHS_COS_WEIGHT + commonScore * self.CHS_COM_WEIGHT

    def compareEngAddress(self, rawAddress1, rawAddress2):
        escapeChars = ['no.', '-', '—', '.', '。', ',', '，', 
            'nt', 'kln', 'kl', 'hk', 'hong kong', 'kowloon', 'new territories',
            '(', ')', '（', '）', '[', ']', '〔', '〕', '{', '}', '｛', '｝', '|', '｜', '+', '＋',
            '/', '／', '\\', '＼', '~', '～', ';', '；', '*', '＊', '&', '＆', '@', '＠', '\'', '’', '"', '”',
            '#', '＃', '!', '！', '$', '＄', '%', '％', '<', '＜', '>', '＞', '^', '︿', '=', '＝', '、']
        address1 = self.preProcessing(rawAddress1, rawAddress2, escapeChars)[0]
        address2 = self.preProcessing(rawAddress1, rawAddress2, escapeChars)[1]

        addr1 = re.findall('\d+', address1)
        addr2 = re.findall('\d+', address2)
        addr1.extend(re.findall('[a-zA-Z]+', address1))
        addr2.extend(re.findall('[a-zA-Z]+', address2))
        totalCos = 0
        totalCom = 0
        for addr1Word in addr1:
            maxCom = 0
            maxCos = 0
            for addr2Word in addr2:
                maxCos = max(maxCos, self.cosSimilarity(addr1Word, addr2Word))
                maxCom = max(maxCom, self.comSimilarity(addr1Word, addr2Word, self.ENG_THRESHOLD))
            totalCos += maxCos
            totalCom += maxCom
        
        cosScore = totalCos / len(addr1)
        commonScore = totalCom / len(addr1)

        return cosScore * self.ENG_COS_WEIGHT + commonScore * self.ENG_COM_WEIGHT

File: AddressSimilarity.cpp
#include <bits/stdc++.h>
#include "ASutils.h"
#include "AddressSimilarity.h"

using namespace std;

// Private

/*--------------------------------------------------------------------------------------------------
 Pre-procoessing two given addresses
 @return: array of string -> Two processed addresses
 @params: string addr1 -> Address 1
          string addr2 -> Address 2
          (Optional) vector<string> escape_chars -> List of characters to be escaped
          (Optional) vector<pair<string, string>> replace_chars -> List of characters to be replaced
--------------------------------------------------------------------------------------------------*/
string *pre_processing(string addr1, string addr2, const vector<string> &escape_chars = {}, const vector<pair<string, string>> &replace_chars = {}) {
  strip(addr1), strip(addr2);
  lower(addr1), lower(addr2);
  for (auto e : ESCAPE_CHARS) {
    replace(addr1, e, "");
    replace(addr2, e, "");
  }
  for (auto e : escape_chars) {
    replace(addr1, e, "");
    replace(addr2, e, "");
  }
  for (auto r : replace_chars) {
    replace(addr1, r.first, r.second);
    replace(addr2, r.first, r.second);
  }
  string *res = new string[2];
  res[0] = addr1, res[1] = addr2;
  return res;
}

/*--------------------------------------------------------------------------------------------------
 Split the given English address
 @return: vector of string -> Vector of splitted words
 @params: string addr -> English address
--------------------------------------------------------------------------------------------------*/
vector<string> split_processing(const string &addr) {
  vector<string> tokens;
  regex rx_d(R"(([[:digit:]]+))");
  smatch sm;
  string str = addr;
  while (regex_search(str, sm, rx_d)) {
    tokens.push_back(sm[0]);
    str = sm.suffix().str();
  }
  regex rx_s(R"(([a-zA-Z]+))");
  str = addr;
  while (regex_search(str, sm, rx_s)) {
    tokens.push_back(sm[0]);
    str = sm.suffix().str();
  }
  return tokens;
}

/*--------------------------------------------------------------------------------------------------
 Compute the consine similarity
 @return: float -> Score of the consine similarity between two addresses
 @params: string raw_addr1 -> Address 1
          string raw_addr2 -> Address 2
--------------------------------------------------------------------------------------------------*/
float cos_similarity(string raw_addr1, string raw_addr2) {
  if (raw_addr1 == raw_addr2) {
    return 1.0;
  }
  if (raw_addr1.size() == 0 || raw_addr2.size() == 0) {
    return 0;
  }
  set<wchar_t> addr1_set, addr2_set, addr_set;
  set<wchar_t>::iterator itr;
  wstring addr1 = utf8_to_wstring(raw_addr1);
  wstring addr2 = utf8_to_wstring(raw_addr2);
  for (wchar_t c : addr1) {
    addr1_set.insert(c);
  }
  for (wchar_t c : addr2) {
    addr2_set.insert(c);
  }
  set_union(addr1_set.begin(), addr1_set.end(), addr2_set.begin(), addr2_set.end(), inserter(addr_set, addr_set.end()));
  vector<float> addr1_count, addr2_count;
  for (auto itr : addr_set) {
    addr1_count.push_back(count(addr1.begin(), addr1.end(), itr));
    addr2_count.push_back(count(addr2.begin(), addr2.end(), itr));
  }
  float addr1_sum = 0, addr2_sum = 0, cross_sum = 0;
  for (int i = 0; i < addr1_count.size(); i++) {
    cross_sum += addr1_count[i] * addr2_count[i];
    addr1_sum += addr1_count[i] * addr1_count[i];
    addr2_sum += addr2_count[i] * addr2_count[i];
  }
  float diff = sqrt(addr1_sum) * sqrt(addr2_sum);
  return diff == 0 ? 0 : cross_sum / diff;
}

/*--------------------------------------------------------------------------------------------------
 Compute the non-overlapping common substrings similarity
 @return: float -> Score of the non-overlapping common substring similarity between two addresses
 @params: string raw_addr1 -> Address 1
          string raw_addr2 -> Address 2
          int threshold -> Threshold of non-overlapping common substrings
--------------------------------------------------------------------------------------------------*/
float com_similarity(string raw_addr1, string raw_addr2, int threshold) {
  if (raw_addr1 == raw_addr2) {
    return 1.0;
  }
  if (raw_addr1.size() == 0 || raw_addr2.size() == 0) {
    return 0;
  }
  wstring addr1 = utf8_to_wstring(raw_addr1);
  wstring addr2 = utf8_to_wstring(raw_addr2);
  float val[addr1.size() + 1][addr2.size() + 1];
  for (int i = 0; i <= addr1.size(); i++) {
    for (int j = 0; j <= addr2.size(); j++) {
      val[i][j] = 0;
    }
  }
  for (int i = 1; i <= addr1.size(); i++) {
    for (int j = 1; j <= addr2.size(); j++) {
      if (addr1[i - 1] == addr2[j - 1]) {
        val[i][j] = val[i - 1][j - 1] + 1;
      }    
    }
  }
  float score = 0;
  int i = addr1.size();
  int previous = threshold != 1;
  while (i > 0) {
    int j = addr2.size();
    while (j > 0) {
      if (val[i][j] >= threshold && val[i - 1][j - 1] >= previous) {
        score += val[i][j];
        i -= val[i][j] - 1;
        break;
      }
      j--;
    }
    i--;
  }
  return score / max(addr1.size(), addr2.size());
}

// Public

/*--------------------------------------------------------------------------------------------------
 AddressSimilarity constructor
 @return: instantiate AddressSimilarity
 @params: float chs_cos_weight -> Default 0.5, Weight for cosine similarity of Chinese address
          float chs_com_weight -> Default 0.5, Weight for non-overlapping common substrings of Chinese address
          int chs_threshold -> Default 2, Threshold of non-overlapping common substrings of Chinese address (at least 1)
          float eng_cos_weight -> Default 0.3, Weight for cosine similarity of English address
          float eng_com_weight -> Default 0.7, Weight for non-overlapping common substrings of English address
          int eng_threshold -> Default 3, Threshold of non-overlapping common substrings of English address (at least 1)
--------------------------------------------------------------------------------------------------*/
AddressSimilarity :: AddressSimilarity(float chs_cos_weight, float chs_com_weight, int chs_threshold,
                                        float eng_cos_weight, float eng_com_weight, int eng_threshold) {
  fprintf(stderr, "AddressSimilarity CREATED\n");
  this -> chs_cos_weight = chs_cos_weight;
  this -> chs_com_weight = chs_com_weight;
  this -> chs_threshold = chs_threshold ? chs_threshold : 1;
  this -> eng_cos_weight = eng_cos_weight;
  this -> eng_com_weight = eng_com_weight;
  this -> eng_threshold = eng_threshold ? eng_threshold : 1;
  srand(time(NULL));
}

/*--------------------------------------------------------------------------------------------------
 AddressSimilarity destructor
--------------------------------------------------------------------------------------------------*/
AddressSimilarity :: ~AddressSimilarity() {
  fprintf(stderr, "AddressSimilarity DESTROYED\n");
}

/*--------------------------------------------------------------------------------------------------
 Function to compare two addresses by auto-detecting language 
 @return: float -> Overall similarity
 @params: string raw_addr1 -> Address 1
          string raw_addr2 -> Address 2
          (Optional) bool enable_rand -> Default true, Enable randomized computation
--------------------------------------------------------------------------------------------------*/
float AddressSimilarity :: compare_addr(string raw_addr1, string raw_addr2, bool enable_rand) {
  if (raw_addr1 == raw_addr2) {
    return 1;
  }
  if (raw_addr1.size() == 0 || raw_addr2.size() == 0) {
    return 0;
  }
  string tmp_addr1 = raw_addr1;
  string tmp_addr2 = raw_addr2;
  strip(tmp_addr1), strip(tmp_addr2);
  for (auto e : ESCAPE_CHARS) {
    replace(tmp_addr1, e, "");
    replace(tmp_addr2, e, "");
  }
  int d_addr1_count = 0, d_addr2_count = 0;
  int a_addr1_count = 0, a_addr2_count = 0;
  for (auto c : tmp_addr1) {
    if (c - '0' >= 0 && c - '0' <= 9) {
      d_addr1_count++;
    } else if ((c - 'a' >= 0 && c - 'a' <= 26) || (c - 'A' >= 0 && c - 'A' <= 26)) {
      a_addr1_count++;
    }
  }
  for (auto c : tmp_addr2) {
    if (c - '0' >= 0 && c - '0' <= 9) {
      d_addr2_count++;
    } else if ((c - 'a' >= 0 && c - 'a' <= 26) || (c - 'A' >= 0 && c - 'A' <= 26)) {
      a_addr2_count++;
    }
  }
  bool addr1_is_chs = tmp_addr1.size() - d_addr1_count > a_addr1_count * 2;
  bool addr2_is_chs = tmp_addr2.size() - d_addr2_count > a_addr2_count * 2;
  if (addr1_is_chs && addr2_is_chs) {
    return compare_chs_addr(raw_addr1, raw_addr2);
  }
  if (!addr1_is_chs && !addr2_is_chs) {
    return compare_eng_addr(raw_addr1, raw_addr2, enable_rand);
  }
  return -1;
}

/*--------------------------------------------------------------------------------------------------
 Function to compare two Chinese addresses
 @return: float -> Overall similarity
 @params: string raw_addr1 -> Address 1
          string raw_addr2 -> Address 2
--------------------------------------------------------------------------------------------------*/
float AddressSimilarity :: compare_chs_addr(string raw_addr1, string raw_addr2) {
  vector<string> escape_chars = {" ", "　", ",", "，", ".", "。",
                                  "號", "号", "鋪", "铺", "舖", "九龍", "九龙", "香港", "新界"};
  vector<pair<string, string>> replace_chars = {{"-", "至"}, {"—", "至"}, {"o", "零"}, {"ｏ", "零"}, {"Ｏ", "零"}, {"&", "及"}, {"＆", "及"}};
  string addr1 = pre_processing(raw_addr1, raw_addr2, escape_chars, replace_chars)[0];
  string addr2 = pre_processing(raw_addr1, raw_addr2, escape_chars, replace_chars)[1];
  if (addr1.size() == 0 || addr2.size() == 0) {
    return 0;
  }
  addr1 = to_chs_num(addr1, addr2);
  addr2 = to_chs_num(addr2, addr1);
  float cos_score = cos_similarity(addr1, addr2);
  float com_score = com_similarity(addr1, addr2, chs_threshold);
  return cos_score * chs_cos_weight + com_score * chs_com_weight;
}

/*--------------------------------------------------------------------------------------------------
 Function to compare two English addresses
 @return: float -> Overall similarity
 @params: string raw_addr1 -> Address 1
          string raw_addr2 -> Address 2
          (Optional) bool enable_rand -> Default true, Enable randomized computation
--------------------------------------------------------------------------------------------------*/
float AddressSimilarity :: compare_eng_addr(string raw_addr1, string raw_addr2, bool enable_rand) {
  vector<string> escape_chars = {"no.", "-", "—",
                                  " nt", " kln", " kl", "hk", "hong kong", "kowloon", "new territories"};
  string addr1 = pre_processing(raw_addr1, raw_addr2, escape_chars)[0];
  string addr2 = pre_processing(raw_addr1, raw_addr2, escape_chars)[1];
  if (addr1.size() == 0 || addr2.size() == 0) {
    return 0;
  }
  vector<string> addr1_tokens = split_processing(addr1);
  vector<string> addr2_tokens = split_processing(addr2);
  float total_cos = 0, total_com = 0;
  for (string s : addr1_tokens) {
    float max_cos = 0, max_com = 0;
    for (string ss : addr2_tokens) {
      max_cos = max(max_cos, cos_similarity(s, ss));
      max_com = max(max_com, com_similarity(s, ss, eng_threshold));
    }
    total_cos += max_cos;
    total_com += max_com;
  }
  float ratio = 1 - min(addr1_tokens.size(), addr2_tokens.size()) * 1.0 / max(addr1_tokens.size(), addr2_tokens.size());
  int offset = (int) (ratio * addr1_tokens.size());
  float cos_score = total_cos / (addr1_tokens.size() + (enable_rand * (rand() % 2) * offset));
  float com_score = total_com / (addr1_tokens.size() + (enable_rand * (rand() % 2) * offset));
  return cos_score * eng_cos_weight + com_score * eng_com_weight;
}

/*--------------------------------------------------------------------------------------------------
 Show current settings of this AddressSimilarity object
 @return: None
 @params: None
--------------------------------------------------------------------------------------------------*/
void AddressSimilarity :: get_params() {
  fprintf(stderr, "CHINESE -> cos weight: %.4f, com weight: %.4f, threshold: %d\n", chs_cos_weight, chs_com_weight, chs_threshold);
  fprintf(stderr, "ENGLISH -> cos weight: %.4f, com weight: %.4f, threshold: %d\n", eng_cos_weight, eng_com_weight, eng_threshold);
}

/*--------------------------------------------------------------------------------------------------
 Edit settings of this AddressSimilarity object
 @return: None
 @params: float chs_cos_weight -> Default 0.5, Weight for cosine similarity of Chinese address
          float chs_com_weight -> Default 0.5, Weight for non-overlapping common substrings of Chinese address
          int chs_threshold -> Default 2, Threshold of non-overlapping common substrings of Chinese address (at least 1)
          float eng_cos_weight -> Default 0.3, Weight for cosine similarity of English address
          float eng_com_weight -> Default 0.7, Weight for non-overlapping common substrings of English address
          int eng_threshold -> Default 3, Threshold of non-overlapping common substrings of English address (at least 1)
--------------------------------------------------------------------------------------------------*/
void AddressSimilarity :: set_params(float chs_cos_weight, float chs_com_weight, int chs_threshold,
                                      float eng_cos_weight, float eng_com_weight, int eng_threshold) {
  this -> chs_cos_weight = chs_cos_weight;
  this -> chs_com_weight = chs_com_weight;
  this -> chs_threshold = chs_threshold ? chs_threshold : 1;
  this -> eng_cos_weight = eng_cos_weight;
  this -> eng_com_weight = eng_com_weight;
  this -> eng_threshold = eng_threshold ? eng_threshold : 1;
}